File: jetbrains.py
import json
import requests
from common import Storage

storage = Storage('jetbrains')

user_agent = 'Mozilla/5.0 (Macintosh; ' \
             'Intel Mac OS X 10_15_7) ' \
             'AppleWebKit/537.36 (KHTML, like Gecko) ' \
             'Chrome/115.0.0.0 Safari/537.36'


def fetch(__code__: str):
    req = requests.get(
        url="https://data.services.jetbrains.com/products?code={}&release.type={}&fields={}".format(
            __code__,
            'eap%2Crc%2Crelease',
            'distributions%2Clink%2Cname%2Creleases'
        ),
        headers={
            'Usage-Agent': user_agent
        }
    )
    obj = json.loads(req.text)
    for o in obj:
        name = o['name']
        for release in o['releases']:
            type_name = release['type']
            if type_name != 'release':
                continue
            major_version = release['majorVersion']
            version = release['version']
            for arch in release['downloads']:
                download = release['downloads'][arch]
                origin_link = download['link']
                s = origin_link.split('/')
                file_name = s[len(s) - 1]
                if len(file_name) > 5 and file_name[len(file_name) - 5:] == ".json":
                    continue
                link = '{}/{}/{}/{}/{}/{}'.format(name, type_name, major_version, version, arch, file_name)
                storage.write(link, origin_link)


def main():
    fetch('CL')  # CLion
    fetch('DG')  # DataGrip
    fetch('DS')  # DataSpell
    fetch('GO')  # GoLand
    fetch('PC')  # PyCharm
    fetch('PS')  # PhpStorm
    fetch('IIU')  # IntelliJ IDEA Ultimate
    fetch('RM')  # RubyMine
    fetch('WS')  # WebStorm


if __name__ == '__main__':
    main()

File: app/data/task/jetbrains.go
package task

import (
	"fmt"
	"strings"

	"github.com/ismdeep/mirror-data/app/data/conf"
	"github.com/ismdeep/mirror-data/app/data/global"
	"github.com/ismdeep/mirror-data/app/data/internal/httputil"
	"github.com/ismdeep/mirror-data/app/data/internal/store"
)

type JetBrains struct {
	storage *store.Storage
}

func (receiver *JetBrains) GetBucketName() string {
	return "jetbrains"
}

func (receiver *JetBrains) Run() {
	receiver.storage = store.New("jetbrains", conf.Config.StorageCoroutineSize)

	defer func() {
		close(receiver.storage.C)
		receiver.storage.WG.Wait()
	}()

	products := map[string]string{
		"CL":  "CLion",
		"DG":  "DataGrip",
		"DS":  "DataSpell",
		"GO":  "GoLand",
		"IIU": "IDEA",
		"PC":  "PyCharm",
		"PCC": "PyCharmCommunity",
		"PS":  "PhpStorm",
		"RD":  "Rider",
		"RM":  "RubyMine",
		"WS":  "WebStorm",
	}

	for code, productName := range products {
		if err := receiver.fetch(code, productName); err != nil {
			global.Errors <- err
			return
		}
	}
}
func (receiver *JetBrains) fetch(code string, productName string) error {
	var products []Product
	reqURL := fmt.Sprintf("https://data.services.jetbrains.com/products?code=%v", code)
	if err := httputil.GetWithJSONUnmarshal(reqURL, &products); err != nil {
		return err
	}

	for _, product := range products {
		for _, release := range product.Releases {
			if release.Type != "release" {
				continue
			}
			for arch, download := range release.Downloads {
				originLink := download.Link
				fileName := originLink[strings.LastIndex(originLink, "/")+1:]
				if len(fileName) <= 5 || fileName[len(fileName)-5:] == ".json" {
					continue
				}
				link := fmt.Sprintf("%v/%v/%v/%v/%v/%v", productName, release.Type, release.MajorVersion, release.Version, arch, fileName)
				receiver.storage.Add(link, originLink)
			}
		}
	}

	return nil
}

// Product model
type Product struct {
	Name     string           `json:"name"`
	Link     string           `json:"link"`
	Releases []ProductRelease `json:"releases"`
}

// ProductRelease model
type ProductRelease struct {
	Build        string `json:"build"`
	Date         string `json:"date"`
	Downloads    map[string]ProductDownload
	MajorVersion string `json:"majorVersion"`
	Type         string `json:"type"`
	Version      string `json:"version"`
}

// ProductDownload download model
type ProductDownload struct {
	ChecksumLink string `json:"checksumLink"`
	Link         string `json:"link"`
	Size         int64  `json:"size"`
}
